ports/raspberrypi/bindings/rp2pio/StateMachine.c

```diff
@@ -618,23 +619,166 @@ const mp_obj_property_t rp2pio_statemachine_writing_obj = {
+              MP_ROM_NONE,
+              MP_ROM_NONE},
+};
+
+
+// =================================================================================================================================
+
+//|     def background_read(
+//|         self,
+//|         once: Optional[WriteableBuffer] = None,
+//|         *,
+//|         loop: Optional[WriteableBuffer] = None,
+//|         swap: bool = False,
+//|     ) -> None:
+//|         """Read data from the RX fifo in the background, with optional looping.
+//|
+//|         First, if any previous ``once`` or ``loop`` buffer has not been started, this function blocks until they have been started.
+//|         This means that any ``once`` or ``loop`` buffer will be read at least once.
+//|         Then the ``once`` and/or ``loop`` buffers are queued. and the function returns.
+//|         The ``once`` buffer (if specified) will be read just once.
+//|         Finally, the ``loop`` buffer (if specified) will continue being read indefinitely.
+//|
+//|         Reads from the FIFO will match the input buffer's element size. For example, bytearray elements
+//|         will perform 8 bit reads from the PIO FIFO. The RP2040's memory bus will duplicate the value into
+//|         the other byte positions. So, pulling more data in the PIO assembly will read the duplicated values.
+//|
+//|         To perform 16 or 32 bits reads from the FIFO use an `array.array` with a type code of the desired
+//|         size, or use `memoryview.cast` to change the interpretation of an
+//|         existing buffer.  To receive just part of a larger buffer, slice a `memoryview`
+//|         of it.
+//|
+//|         Most use cases will probably only use one of ``once`` or ``loop``.
+//|
+//|         Having neither ``once`` nor ``loop`` terminates an existing
+//|         background looping read after exactly a whole loop. This is in contrast to
+//|         `stop_background_read`, which interrupts an ongoing DMA operation.
+//|
+//|         :param ~Optional[circuitpython_typing.WriteableBuffer] once: Data to be read once
+//|         :param ~Optional[circuitpython_typing.WriteableBuffer] loop: Data to be read repeatedly
+//|         :param bool swap: For 2- and 4-byte elements, swap (reverse) the byte order
+//|         """
+//|         ...
+
+static void fill_buf_info_read(sm_buf_info *info, mp_obj_t obj, size_t *stride_in_bytes) {
+    if (obj != mp_const_none) {
+        info->obj = obj;
+        mp_get_buffer_raise(obj, &info->info, MP_BUFFER_WRITE);
+        size_t stride = mp_binary_get_size('@', info->info.typecode, NULL);
+        if (stride > 4) {
+            mp_raise_ValueError(MP_ERROR_TEXT("Buffer elements must be 4 bytes long or less"));
+        }
+        if (*stride_in_bytes && stride != *stride_in_bytes) {
+            mp_raise_ValueError(MP_ERROR_TEXT("Mismatched data size"));
+        }
+        *stride_in_bytes = stride;
+    } else {
+        memset(info, 0, sizeof(*info));
+    }
+}
```

Comment:
Can you please combine the fill_buf_info_{read,write} routines? I notice you've added detection of mismatched stride between the "once" and "loop" variants, which should probably apply to write as well. I think you'd just have to add the access type parameter and pass MP_BUFFER_{READ,WRITE} in.

Or is there a reason the new checking you added should not also have been done for the background write case?